File: backend/transcend/src/test/java/code/users/logic/LoginTest.java
package code.users.logic;

import static code.users.domain.model.UserFixtures.EMAIL_FIXTURE;
import static code.users.domain.model.UserFixtures.HASH_FIXTURE;
import static code.users.domain.model.UserFixtures.PASSWORD_FIXTURE;
import static code.users.domain.model.UserFixtures.TOKEN_FIXTURE;
import static code.users.domain.model.UserFixtures.aDefaultUser;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import code.users.domain.exceptions.InvalidCredentialsException;
import code.users.ports.in.LoginUseCase;
import code.users.ports.in.LoginUseCase.LoginCommand;
import code.users.ports.out.AccessTokenProvider;
import code.users.ports.out.HashingService;
import code.users.ports.out.UserDao;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;

@SpringJUnitConfig(LoginTest.LoginServiceTestConfig.class)
@RequiredArgsConstructor(onConstructor_ = {@Autowired})
class LoginTest {

  @Configuration
  @Import(Login.class)
  static class LoginServiceTestConfig {}

  private final LoginUseCase service;

  @MockitoBean private UserDao userDao;
  @MockitoBean private HashingService hashingService;
  @MockitoBean private AccessTokenProvider accessTokenProvider;

  @Test
  void loginAuthenticatesAndReturnsBearerToken() {
    // given
    var command = new LoginCommand(EMAIL_FIXTURE, PASSWORD_FIXTURE);
    var user = aDefaultUser();

    when(userDao.findByEmail(EMAIL_FIXTURE)).thenReturn(Optional.of(user));
    when(hashingService.matches(PASSWORD_FIXTURE, HASH_FIXTURE)).thenReturn(true);
    when(accessTokenProvider.generateToken(EMAIL_FIXTURE)).thenReturn(TOKEN_FIXTURE);

    // when
    var result = service.login(command);

    // then
    assertEquals(TOKEN_FIXTURE, result.accessToken());
    assertEquals("Bearer", result.tokenType());
    verify(userDao).findByEmail(EMAIL_FIXTURE);
    verify(hashingService).matches(PASSWORD_FIXTURE, HASH_FIXTURE);
    verify(accessTokenProvider).generateToken(EMAIL_FIXTURE);
  }

  @Test
  void loginWithInvalidEmailThrowsException() {
    // given
    var command = new LoginCommand(EMAIL_FIXTURE, PASSWORD_FIXTURE);

    when(userDao.findByEmail(EMAIL_FIXTURE)).thenReturn(Optional.empty());

    // when & then
    assertThrows(InvalidCredentialsException.class, () -> service.login(command));
  }

  @Test
  void loginWithInvalidPasswordThrowsException() {
    // given
    var command = new LoginCommand(EMAIL_FIXTURE, "wrong-password");
    var user = aDefaultUser();
    when(userDao.findByEmail(EMAIL_FIXTURE)).thenReturn(Optional.of(user));
    when(hashingService.matches("wrong-password", PASSWORD_FIXTURE)).thenReturn(false);

    // when & then
    assertThrows(InvalidCredentialsException.class, () -> service.login(command));
  }
}

File: backend/transcend/src/test/kotlin/code/users/logic/LoginTest.kt
package code.users.logic

import code.bootstrap.config.TokenConfig.TOKEN_TYPE
import code.shared.config.DaoTestSupport
import code.users.domain.exceptions.InvalidCredentialsException
import code.users.domain.model.AuthFixtures.TOKEN_FIXTURE
import code.users.domain.model.UserFixtures.*
import code.users.infrastructure.persistence.UserEntityMapperImpl
import code.users.infrastructure.persistence.UserRepository
import code.users.ports.`in`.LoginUseCase
import code.users.ports.`in`.LoginUseCase.LoginCommand
import code.users.ports.out.AccessTokenProvider
import code.users.ports.out.HashingService
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.shouldBe
import org.mockito.BDDMockito.given
import org.springframework.context.annotation.Import
import org.springframework.test.context.bean.override.mockito.MockitoBean

@Import(
  UserRepository::class,
  UserEntityMapperImpl::class,
  Login::class
)
class LoginComponentTest (
  private val service: LoginUseCase
) : DaoTestSupport() {

  @MockitoBean private lateinit var hashingService: HashingService
  @MockitoBean private lateinit var accessTokenProvider: AccessTokenProvider

  init {

    Given("an existing user in the database") {
      val existingUser = aDaoUser()
      userDao.createUser(existingUser)

      And("a login command with the correct password") {
        val command = LoginCommand(EMAIL_FIXTURE, PASSWORD_FIXTURE)

        When("the login service is executed") {
          given(hashingService.matches(PASSWORD_FIXTURE, HASH_FIXTURE)).willReturn(true)
          given(accessTokenProvider.generateToken(UUID_FIXTURE.toString())).willReturn(TOKEN_FIXTURE)

          val result = service.login(command)

          Then("it should return a valid bearer token") {
            result.accessToken() shouldBe TOKEN_FIXTURE
            result.tokenType() shouldBe TOKEN_TYPE
            result.userId() shouldBe UUID_FIXTURE.toString()
          }
        }
      }

      And("a login command with an incorrect password") {
        val command = LoginCommand(EMAIL_FIXTURE, WRONG_PASSWORD_FIXTURE)

        When("the login service is executed") {
          given(hashingService.matches(WRONG_PASSWORD_FIXTURE, HASH_FIXTURE)).willReturn(false)

          Then("it should throw an InvalidCredentialsException") {
            shouldThrow<InvalidCredentialsException> {
              service.login(command)
            }
          }
        }
      }
    }

    Given("user does not exist") {
      val command = LoginCommand("nonexistent@example.com", PASSWORD_FIXTURE)

      When("the login service is executed") {
        Then("it should throw an InvalidCredentialsException") {
          shouldThrow<InvalidCredentialsException> {
            service.login(command)
          }
        }
      }
    }
  }
}